drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt

```diff
@@ -8,191 +8,150 @@
-    private var imageLoader: ImageLoader,
-    private var showInExpandStatus: Boolean,
+    private val paymentMethodsListModel: PaymentMethodsListModel,
+    private val imageLoader: ImageLoader,
     private val onPaymentMethodSelectedCallback: OnPaymentMethodSelectedCallback
 ) : RecyclerView.Adapter<PaymentMethodAdapter.BaseViewHolder>() {
 
-    private var expandListOfItem = emptyList<PaymentMethod>()
-    private var collapseListOfItem = emptyList<PaymentMethod>()
-
-    init {
-        initAdapter()
-    }
-
-    companion object {
-        internal val TAG = LogUtil.getTag()
-
-        internal const val PAYMENT_METHODS_HEADER = 1
-        internal const val PAYMENT_METHODS_FOOTER = 2
-        internal const val PAYMENT_METHOD_ITEMS = 3
-
-        internal val HEADER_PLACEHOLDER_ITEM_SP = PaymentMethod()
-        internal val HEADER_PLACEHOLDER_ITEM_P = PaymentMethod()
-        internal val FOOTER_PLACEHOLDER_ITEM = PaymentMethod()
-
-        private const val CARD_LOGO_TYPE = "card"
-    }
+    private val headerCount =
+        if (paymentMethodsListModel.storedPaymentMethods.isEmpty()) 1
+        else 2
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
         return when (viewType) {
             PAYMENT_METHODS_HEADER -> PaymentMethodsHeaderVH(getView(parent, R.layout.payment_methods_list_header))
-            PAYMENT_METHODS_FOOTER -> PaymentMethodsFooterVH(getView(parent, R.layout.payment_methods_list_footer))
-            else -> PaymentMethodVH(getView(parent, R.layout.payment_methods_list_item))
+            STORED_PAYMENT_METHOD -> StoredPaymentMethodVH(getView(parent, R.layout.payment_methods_list_item))
+            PAYMENT_METHOD -> PaymentMethodVH(getView(parent, R.layout.payment_methods_list_item))
+            else -> throw CheckoutException("Unexpected viewType on onCreateViewHolder - $viewType")
         }
     }
 
     override fun getItemViewType(position: Int): Int {
-        return when (getPaymentMethod(position)) {
-            HEADER_PLACEHOLDER_ITEM_P, HEADER_PLACEHOLDER_ITEM_SP -> PAYMENT_METHODS_HEADER
-            FOOTER_PLACEHOLDER_ITEM -> PAYMENT_METHODS_FOOTER
-            else -> PAYMENT_METHOD_ITEMS
+        // Top is always a header
+        return if (position == 0) {
+            PAYMENT_METHODS_HEADER
+        } else {
+            if (noStored()) {
+                // only regular payment methods left
+                PAYMENT_METHOD
+            } else {
+                when {
+                    // 0 header - stored payments - 2nd header - regular payments
+                    position <= paymentMethodsListModel.storedPaymentMethods.size -> STORED_PAYMENT_METHOD
+                    position == (paymentMethodsListModel.storedPaymentMethods.size + 1) -> PAYMENT_METHODS_HEADER
+                    else -> PAYMENT_METHOD
+                }
+            }
         }
     }
 
     override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
```

Comment:
Can we extract 3 methods from this? Something like `bindPaymentMethodsHeaderVH` - `bindStoredPaymentMethodVH` - `bindPaymentMethodVH` ?

Reply:
yeah I was considering that too, thanks